src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshalling.cs

```diff
@@ -127,32 +138,237 @@ internal sealed class DisableRuntimeMarshallingAnalyzer : DiagnosticAnalyzer
+            private readonly IMethodSymbol? _getDelegateForFunctionPointerGeneric;
+            private readonly IMethodSymbol? _getFunctionPointerForDelegateNonGeneric;
+            private readonly IMethodSymbol? _getFunctionPointerForDelegateGeneric;
+            private readonly AutoLayoutTypeCache _autoLayoutCache;
+
+            public DelegateInteropUsageAnalyzer(Compilation compilation, AutoLayoutTypeCache autoLayoutCache)
+            {
+                if (compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeInteropServicesMarshal, out var marshalType))
+                {
+                    var getDelegateForFunctionPointerMethods = marshalType.GetMembers("GetDelegateForFunctionPointer").OfType<IMethodSymbol>();
+                    _getDelegateForFunctionPointerNonGeneric = getDelegateForFunctionPointerMethods.First(m => m.TypeArguments.Length == 0);
+                    _getDelegateForFunctionPointerGeneric = getDelegateForFunctionPointerMethods.First(m => m.TypeArguments.Length != 0);
+                    var getFunctionPointerForDelegateMethods = marshalType.GetMembers("GetFunctionPointerForDelegate").OfType<IMethodSymbol>();
+                    _getFunctionPointerForDelegateNonGeneric = getFunctionPointerForDelegateMethods.First(m => m.TypeArguments.Length == 0);
+                    _getFunctionPointerForDelegateGeneric = getFunctionPointerForDelegateMethods.First(m => m.TypeArguments.Length != 0);
+                }
+
+                _autoLayoutCache = autoLayoutCache;
+            }
+
+            public void RegisterActions(CompilationStartAnalysisContext context, bool hasDisableRuntimeMarshallingAttribute)
+            {
+                context.RegisterOperationAction(AnalyzeInvocation, OperationKind.Invocation);
+
+                if (!hasDisableRuntimeMarshallingAttribute)
+                {
+                    context.RegisterSymbolAction(AnalyzeMethod, SymbolKind.Method);
+
+                    context.RegisterOperationAction(AnalyzeLocalFunction, OperationKind.LocalFunction);
+                }
+            }
+
+            private void AnalyzeInvocation(OperationAnalysisContext context)
+            {
+                IInvocationOperation operation = (IInvocationOperation)context.Operation;
+                if (SymbolEqualityComparer.Default.Equals(operation.TargetMethod.OriginalDefinition, _getDelegateForFunctionPointerNonGeneric))
+                {
+                    if (operation.Arguments.Length == 2 && operation.Arguments[1] is { Value: ITypeOfOperation { TypeOperand: INamedTypeSymbol { TypeKind: TypeKind.Delegate } delegateType } })
+                    {
+                        AnalyzeDelegateType(context, operation.Arguments[1], delegateType);
+                    }
+                }
+                else if (SymbolEqualityComparer.Default.Equals(operation.TargetMethod.OriginalDefinition, _getDelegateForFunctionPointerGeneric))
```

Comment:
FYI: I am not reviewing the correctness/exhaustiveness of your analysis here, I don't think I have the expertise or knowledge for it.